### PR TITLE
Fix memory leak reported by ASAN in NNVM to ONNX conversion

### PR DESCRIPTION
**Pull request by kevinzh92**
## Description ##
Fix memory leak reported by ASAN in NNVM to ONNX conversion of a constant.

### Essentials ###
Please feel free to remove inapplicable items for your PR.
- [ ] The PR title starts with [MXNET-$JIRA_ID], where $JIRA_ID refers to the relevant [JIRA issue](https://issues.apache.org/jira/projects/MXNET/issues) created (except PRs with tiny changes)
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)
- Build tests will be added for build configuration changes (e.g. adding a new build option with NCCL)
- [ ] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- Check the API doc at http://mxnet-ci-doc.s3-accelerate.dualstack.amazonaws.com/PR-$PR_ID/$BUILD_ID/index.html
- [x] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
- [x] Fixed a memory leak in NNVM to ONNX constants conversion code. The `shared_ptr`, as written, does not call `delete[]` on a dynamically allocated array.

## Comments ##

### CODE REVIEW

**Comment by ChaiBapchya** [comment]
@larroy isn't it consistent with the existing naming system?  
That is, to call the instance of that class as the corresponding data pointer.

**Comment by larroy** [comment]
Not really, in those cases you want a semantic name, even if you are wrapping your Cat in a pointer you want to name it:

unique_ptr<Cat> cat = make_unique<Cat>() 

And not 

unique_ptr<Cat> unique_ptr = make_unique<Cat>().

Semantic naming is important, we have already type information from the compiler and we are in a statically typed language, no need to name a pointer _ptr. And definitely not "data" for the name of a variable, this is a code smell.

**Comment by ChaiBapchya** [comment]
Gotcha! Makes sense!

**Comment by karan6181** [comment]
@kevinzh92 Could you please address the review comments?

**Comment by kevinzh92** [comment]
Sorry guys. Things have been busy at work and this fell off my radar. @larroy, I hope I addressed all of your concerns.

File: src/operator/subgraph/tensorrt/nnvm_to_onnx.cc [hunk 1]
```diff
@@ -519,21 +519,18 @@ void ConvertConstant(
...
   if (dtype == TensorProto_DataType_FLOAT) {
-    std::shared_ptr<float> shared_data_ptr(new float[size]);
-    float* const data_ptr = shared_data_ptr.get();
-    nd.SyncCopyToCPU(static_cast<void*>(data_ptr), size);
-
-    for (size_t blob_idx = 0; blob_idx < size; ++blob_idx) {
-      initializer_proto->add_float_data(data_ptr[blob_idx]);
-    }
+      std::vector<float> constants(size);
+      nd.SyncCopyToCPU(&constants, size);
```
Comment (larroy) [inline src/operator/subgraph/tensorrt/nnvm_to_onnx.cc]: 🙅‍♂️this is not correct. You are taking address to the vector itself, you want to take address to the data.

Please use .data() member or the address of the first element.

**Comment by KellenSunderland** [comment]
Looks like there's still a compiler issue:
```
/work/mxnet/src/operator/subgraph/tensorrt/nnvm_to_onnx.cc:567:54: error: reinterpret_cast from type 'const short unsigned int*' to type 'int32_t* {aka int*}' casts away qualifiers

                   reinterpret_cast<int32_t&>(constant));
```

**Comment by larroy** [comment]
I also have concerns about this line:

```
                  reinterpret_cast<int32_t&>(constant));
```

**Comment by wkcn** [comment]
Hi, I update this code to use `T[]` as the template type of `shared_ptr`.

Thank @kevinzh92 for pointing out this issue.

**Comment by larroy** [comment]
Please rebase

**Comment by wkcn** [comment]
Merged.
It is a tiny change to change the template type, and keep the original variable name.
We can modify the semantic name in another PR.
std::vector is not suitable in this PR, since it will allocate extra capacity.

Thank you!

**Pull request merged**